dxr/plugins/descriptor/__init__.py

```diff
@@ -0,0 +1,110 @@
+"""Descriptor - Provide useful descriptions for common file types.
+Refer to https://mxr.mozilla.org/webtools-central/source/mxr/Local.pm#27
+"""
+import re
+from os import listdir
+from os.path import splitext, basename, join, isfile
+from warnings import warn
+
+import dxr.indexers
+
+
+class FolderToIndex(dxr.indexers.FolderToIndex):
+    browse_headers = ['Description']
+
+    def needles(self):
+        """If the folder contains a readme, then yield the first line of the
+        readme as the description.
+        Similar to https://mxr.mozilla.org/webtools-central/source/mxr/Local.pm#251
+        """
+        for entry in listdir(self.path):
+            path = join(self.path, entry)
+            # If we find a readme, then open it and return the first line if
+            # it's non-empty.
+            if "readme" in entry.lower() and isfile(path):
+                with open(path) as readme:
+                    first_line = readme.readline(100).strip()
+                    if first_line:
+                        # Pack into a list for consistency with the file needle.
+                        return [("Description", [first_line])]
+        # Didn't find anything to use as a description
+        return []
+
+
+class FileToIndex(dxr.indexers.FileToIndex):
+    """Do lots of work to yield a description needle."""
+
+    comment_re = re.compile(r'(?:.*?/\*+)(?:\s*\*?\s*)(?P<description>.*?)(?:(?:\*+/.*)|(?:$))', flags=re.M)
+    docstring_re = re.compile(r'(\'\'\'|""")(?:\s*)(?P<description>.*?)(?:(?:(\'\'\'|"""))|(?:$))', flags=re.M)
```

Comment:
I puzzled my head against the avoiding non-greedy quantifiers, but I'm not sure how else to say "scan until 3 quotes in a row".